src/Analyzer/Passes/FunctionToSubcolumnsPass.cpp

```diff
@@ -16,218 +17,419 @@
+    if (!second_argument_constant_node)
+        return;
+
+    const auto & data_type_concrete = assert_cast<const DataType &>(*ctx.column.type);
+    auto subcolumn_name = getSubcolumnNameForElement(second_argument_constant_node->getValue(), data_type_concrete);
+
+    if (subcolumn_name.empty())
+        return;
+
+    NameAndTypePair column{ctx.column.name + "." + subcolumn_name, function_node.getResultType()};
+    node = std::make_shared<ColumnNode>(column, ctx.column_source);
+}
+
+std::map<std::pair<TypeIndex, String>, NodeToSubcolumnTransformer> node_transformers =
+{
+    {
+        {TypeIndex::Array, "length"}, optimizeFunctionLength,
+    },
+    {
+        {TypeIndex::Array, "empty"}, optimizeFunctionEmpty<true>,
+    },
+    {
+        {TypeIndex::Array, "notEmpty"}, optimizeFunctionEmpty<false>,
+    },
+    {
+        {TypeIndex::Map, "length"}, optimizeFunctionLength,
+    },
+    {
+        {TypeIndex::Map, "empty"}, optimizeFunctionEmpty<true>,
+    },
+    {
+        {TypeIndex::Map, "notEmpty"}, optimizeFunctionEmpty<false>,
+    },
+    {
+        {TypeIndex::Map, "mapKeys"},
+        [](QueryTreeNodePtr & node, FunctionNode & function_node, ColumnContext & ctx)
+        {
+            /// Replace `mapKeys(map_argument)` with `map_argument.keys`
+            NameAndTypePair column{ctx.column.name + ".keys", function_node.getResultType()};
+            node = std::make_shared<ColumnNode>(column, ctx.column_source);
+        },
+    },
+    {
+        {TypeIndex::Map, "mapValues"},
+        [](QueryTreeNodePtr & node, FunctionNode & function_node, ColumnContext & ctx)
+        {
+            /// Replace `mapValues(map_argument)` with `map_argument.values`
+            NameAndTypePair column{ctx.column.name + ".values", function_node.getResultType()};
+            node = std::make_shared<ColumnNode>(column, ctx.column_source);
+        },
+    },
+    {
+        {TypeIndex::Map, "mapContains"},
+        [](QueryTreeNodePtr &, FunctionNode & function_node, ColumnContext & ctx)
+        {
+            /// Replace `mapContains(map_argument, argument)` with `has(map_argument.keys, argument)`
+            const auto & data_type_map = assert_cast<const DataTypeMap &>(*ctx.column.type);
+
+            NameAndTypePair column{ctx.column.name + ".keys", std::make_shared<DataTypeArray>(data_type_map.getKeyType())};
+            auto & function_arguments_nodes = function_node.getArguments().getNodes();
+
+            auto has_function_argument = std::make_shared<ColumnNode>(column, ctx.column_source);
+            function_arguments_nodes[0] = std::move(has_function_argument);
+
+            resolveOrdinaryFunctionNodeByName(function_node, "has", ctx.context);
+        },
+    },
+    {
+        {TypeIndex::Nullable, "count"},
+        [](QueryTreeNodePtr &, FunctionNode & function_node, ColumnContext & ctx)
+        {
+            /// Replace `count(nullable_argument)` with `sum(not(nullable_argument.null))`
```

Comment:
Should we check that the argument is nullable?

Reply:
I think no. This function is supposed to be called only for nullable types (we have `TypeIndex::Nullable` in the key of map) and we also don't check types in other functions.